Update JWT guide with the builder API intro and fix some typos

sberyozkin opened the pull request:
This PR provides the introduction to the JWT Builder API and also fixes few typos and adds two missing properties (as spotted by a smallrye-jwt committer @MikeEdgar during the smallrye-jwt PR review)

sberyozkin committed:
Update JWT guide with the builder API intro and fix some typos

gastaldi commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
 |smallrye.jwt.token.cookie|none|Name of the cookie containing a token. This property will be effective only if  `smallrye.jwt.token.header` is set to `Cookie`.
 |smallrye.jwt.token.schemes|`Bearer`|Comma-separated list containing an alternative single or multiple schemes, for example, `DPoP`.
+|smallrye.jwt.token.kid|none|Key identifier. If it is set then the verification JWK key as well every JWT token must have a matching 'kid' header.
+|smallrye.jwt.time-to-live|none|The maximum number of seconds that a JWT may be issued for use. Effectively, the difference  between the expiration date of the JWT and the issued at date must not exceed this value.
```
```suggestion
|smallrye.jwt.time-to-live|none|The maximum number of seconds that a JWT may be issued for use. Effectively, the difference between the expiration date of the JWT and the issued at date must not exceed this value.
```

gastaldi commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+== Generate JWT tokens with Smallrye JWT
+
+JWT claims can be signed or encrypted or signed first and the nested JWT token - encrypted.
+Signing the claims is used most ofthen to secure the claims. What is known today as a JWT token is typically produced by signing the claims in a JSON format using the steps described in the `JSON Web Signature RFC7515` specification.
```
```suggestion
Signing the claims is used most often to secure the claims. What is known today as a JWT token is typically produced by signing the claims in a JSON format using the steps described in the `JSON Web Signature RFC7515` specification.
```

gastaldi commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+
+JWT claims can be signed or encrypted or signed first and the nested JWT token - encrypted.
+Signing the claims is used most ofthen to secure the claims. What is known today as a JWT token is typically produced by signing the claims in a JSON format using the steps described in the `JSON Web Signature RFC7515` specification.
+However, when the claims are sensitive, their confidenatiality can be guaranteed by following the steps described in the `JSON Web Encryption RFC7516` specification to produce a JWT token with the encrypted claims.
```
```suggestion
However, when the claims are sensitive, their confidentiality can be guaranteed by following the steps described in the `JSON Web Encryption RFC7516` specification to produce a JWT token with the encrypted claims.
```

gastaldi replied:
Also a link to the `JSON Web Encryption RFC7516` spec would be nice

gastaldi reviewed (dismissed):
Found some typos, but otherwise looks really good

sberyozkin commented:
@gastaldi Hi, sorry for all those typos, will deal with them, thanks

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -298,7 +298,7 @@ nQIDAQAB
 === Generating a JWT
 
-Often one obtains a JWT from an identity manager like https://www.keycloak.org/[Keycloak], but for this quickstart we will generate our own using the JWT generation API provided by `smallrye-jwt` and the TokenUtils class shown in the following listing. Take this source and place it into `security-jwt-quickstart/src/test/java/org/acme/jwt/TokenUtils.java`.
+Often one obtains a JWT from an identity manager like https://www.keycloak.org/[Keycloak], but for this quickstart we will generate our own using the JWT generation API provided by `smallrye-jwt` (see <<Create JWT tokens with Smallrye JWT>> for more infomation) and the TokenUtils class shown in the following listing. Take this source and place it into `security-jwt-quickstart/src/test/java/org/ac ...
```
Please use an id to make the link. If you change the title, the link will be broken.

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+
+== Generate JWT tokens with Smallrye JWT
+
+JWT claims can be signed or encrypted or signed first and the nested JWT token - encrypted.
```
```suggestion
JWT claims can be signed or encrypted or signed first and the nested JWT token encrypted.
```

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+
+Smallrye JWT provides an API for securing the JWT claims using all of these options.
+
+=== Create JwtClaimsBuilder and set the claims.
```
```suggestion
=== Create JwtClaimsBuilder and set the claims
```

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+However, when the claims are sensitive, their confidenatiality can be guaranteed by following the steps described in the `JSON Web Encryption RFC7516` specification to produce a JWT token with the encrypted claims.
+Finally both the confidentiality and integrity of the claims can be further enforced by signing them first and then encrypting the nested JWT token.
+
+Smallrye JWT provides an API for securing the JWT claims using all of these options.
```
```suggestion
SmallRye JWT provides an API for securing the JWT claims using all of these options.
```

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+
+=== Create JwtClaimsBuilder and set the claims.
+
+The first step is to initialize `JwtClaimsBuilder` using one of the options below and add some claims to it:
```
```suggestion
The first step is to initialize a `JwtClaimsBuilder` using one of the options below and add some claims to it:
```

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+// Builder created from the existing claims
+JwtClaimsBuilder builder2 = Jwt.claims("/tokenClaims.json");
+
+// Builder created from the map of claims
```
```suggestion
// Builder created from a map of claims
```

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+JwtClaimsBuilder builder3 = Jwt.claims(Collections.singletonMap("customClaim", "custom-value"));
+----
+
+The API is fluent so the builder initialization can be done as part of the fluent API sequence. The builder will also set an `iat` (issued at) to the current time, `exp` (expires at) to 5 mins away from the current time and `jti` (unique token identifier) claims if they have not already been set, so one can skip setting them when possible.
```
```suggestion
The API is fluent so the builder initialization can be done as part of the fluent API sequence. The builder will also set `iat` (issued at) to the current time, `exp` (expires at) to 5 minutes away from the current time and `jti` (unique token identifier) claims if they have not already been set, so one can skip setting them when possible.
```

gsmet commented on docs/src/main/asciidoc/security-jwt.adoc:
```diff
@@ -948,13 +948,106 @@ Smallrye JWT provides more properties which can be used to customize the token p
+String jwt2 = Jwt.claims("/tokenClaims.json").jwe().header("custom-header", "custom-value").encrypt(getPublicKey());
+----
+
+Note the `alg` (key management algorithm) header is set to `RSA-OAEP-256` (it will be changed to `RSA-OAEP` in the future version of smallrye-jwt) and the `enc` (content encryption header) is set to `AGCM256` by default.
```
```suggestion
Note the `alg` (key management algorithm) header is set to `RSA-OAEP-256` (it will be changed to `RSA-OAEP` in a future version of smallrye-jwt) and the `enc` (content encryption header) is set to `AGCM256` by default.
```

gsmet reviewed (changes requested):
Nice improvements. I added a few comments. Could you adjust and then squash everything?

Thanks!

sberyozkin commented:
Hi @gsmet thanks, I think I've addressed all the comments from yourself and @gastaldi, added a couple of links too. Have a look please.
One thing I'm not 100% sure about, if it is worth mentioning that this API is going through the MP JWT standardization effort at the moment or not. It will take a few months to settle and I'm not sure when it will be available as a standard API to the Quarkus users. I'd rather update the docs, when it available, and update the forum that a package update is needed (which if all goes well, would be the only change that will be required to migrate). 
What do you all think ?

sberyozkin commented:
I'd rather update the forum after the release and point to the docs and note that in some not too distant future an option of migrating to a standard API will be available...Something like that

gsmet commented:
I think we can work on that incrementally so I would favor the approach you proposed.

gsmet commented:
Merged, thanks!

sberyozkin commented:
@gsmet Thanks :-)
